// xiome/x/features/store/models/submodels/connect-submodel.js
import { pub } from "../../../../toolbox/pub.js";
import { ops } from "../../../../framework/ops.js";
import { makeActivator } from "../utils/make-activator.js";
import { StripeConnectStatus } from "../../types/store-concepts.js";
export function makeConnectSubmodel({ state, allowance, connectService, triggerStripeConnectPopup, }) {
    const change = pub();
    async function loadConnectDetails() {
        if (allowance.connectStripeAccount) {
            await ops.operation({
                promise: connectService.loadConnectDetails(),
                setOp: op => {
                    var _a, _b;
                    state.writable.stripeConnect.connectStatusOp = ops.replaceValue(op, (_a = ops.value(op)) === null || _a === void 0 ? void 0 : _a.connectStatus);
                    state.writable.stripeConnect.connectDetailsOp = ops.replaceValue(op, (_b = ops.value(op)) === null || _b === void 0 ? void 0 : _b.connectDetails);
                },
            });
        }
        else if (allowance.manageStore) {
            await ops.operation({
                promise: connectService.loadConnectStatus(),
                setOp: op => state.writable.stripeConnect.connectStatusOp = op,
            });
        }
    }
    const activator = makeActivator(loadConnectDetails);
    return {
        activate: activator.activate,
        refresh: activator.refreshIfActivated,
        onChange: change.subscribe,
        async connectStripeAccount() {
            await triggerStripeConnectPopup(await connectService.generateConnectSetupLink());
            await loadConnectDetails();
            await change.publish();
        },
        async generateStripeLoginLink() {
            if (ops.value(state.readable.stripeConnect.connectStatusOp) === StripeConnectStatus.Unlinked)
                throw new Error("no stripe account to generate login link for");
            return connectService.generateStripeLoginLink();
        },
        async pause() {
            await connectService.pause();
            state.writable.stripeConnect.connectStatusOp = ops.ready(StripeConnectStatus.Paused);
            if (allowance.manageStore)
                state.writable.stripeConnect.connectDetailsOp = ops.ready({
                    ...ops.value(state.readable.stripeConnect.connectDetailsOp),
                    paused: true,
                });
        },
        async resume() {
            await connectService.resume();
            state.writable.stripeConnect.connectStatusOp = ops.ready(StripeConnectStatus.Ready);
            if (allowance.manageStore)
                state.writable.stripeConnect.connectDetailsOp = ops.ready({
                    ...ops.value(state.readable.stripeConnect.connectDetailsOp),
                    paused: false,
                });
        },
    };
}
