// xiome/x/features/store/models/submodels/subscription-planning-submodel.js
import { substate } from "@chasemoskal/snapstate";
import { ops } from "../../../../framework/ops.js";
import { makeActivator } from "../utils/make-activator.js";
export function makeSubscriptionPlanningSubmodel({ state, allowance, subscriptionPlanningService, }) {
    const planningState = substate(state, tree => tree.subscriptionPlanning);
    async function loadSubscriptionPlans() {
        await ops.operation({
            setOp: op => planningState.writable.subscriptionPlansOp = op,
            promise: subscriptionPlanningService.listSubscriptionPlans(),
        });
    }
    const activator = makeActivator(loadSubscriptionPlans);
    return {
        activate: activator.activate,
        refresh: activator.refreshIfActivated,
        async addPlan(options) {
            var _a;
            const newPlan = await subscriptionPlanningService.addPlan(options);
            const oldPlans = (_a = ops.value(planningState.readable.subscriptionPlansOp)) !== null && _a !== void 0 ? _a : [];
            planningState.writable.subscriptionPlansOp = ops.replaceValue(planningState.readable.subscriptionPlansOp, [...oldPlans, newPlan]);
            return newPlan;
        },
    };
}
