// @benev/slate/x/shiny/units/carbon.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
import { css } from "lit";
import { GoldElement } from "../../element/gold.js";
import { UseCarbon } from "../parts/use/tailored.js";
import { setup_reactivity } from "../parts/setup_reactivity.js";
export const prepare_carbon = ((shell) => (settings, renderer) => {
    var _use, _rend, _reactivity, _a;
    return _a = class extends GoldElement {
            constructor() {
                super(...arguments);
                _use.set(this, new UseCarbon(this, this.root, () => void this.requestUpdate(), shell.context));
                _rend.set(this, UseCarbon.wrap(__classPrivateFieldGet(this, _use, "f"), () => renderer(__classPrivateFieldGet(this, _use, "f"))));
                _reactivity.set(this, void 0);
            }
            static get styles() {
                var _b;
                return [
                    shell.context.theme,
                    (_b = settings.styles) !== null && _b !== void 0 ? _b : css ``,
                ];
            }
            render() {
                var _b;
                return (_b = __classPrivateFieldGet(this, _reactivity, "f")) === null || _b === void 0 ? void 0 : _b.render();
            }
            connectedCallback() {
                super.connectedCallback();
                UseCarbon.reconnect(__classPrivateFieldGet(this, _use, "f"));
                __classPrivateFieldSet(this, _reactivity, setup_reactivity(__classPrivateFieldGet(this, _rend, "f"), () => void this.requestUpdate()), "f");
            }
            disconnectedCallback() {
                super.disconnectedCallback();
                UseCarbon.disconnect(__classPrivateFieldGet(this, _use, "f"));
                if (__classPrivateFieldGet(this, _reactivity, "f")) {
                    __classPrivateFieldGet(this, _reactivity, "f").stop();
                    __classPrivateFieldSet(this, _reactivity, undefined, "f");
                }
            }
        },
        _use = new WeakMap(),
        _rend = new WeakMap(),
        _reactivity = new WeakMap(),
        _a.label = settings.name,
        _a;
});
