// @benev/slate/x/nexus/parts/use/parts/use.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Use_context, _Use_rerender, _Use_counter, _Use_mounts, _Use_unmounts, _Use_initStarts, _Use_initResults, _Use_initDowns, _Use_onces, _Use_deferred, _Use_deferredResults, _Use_states, _Use_flatstates, _Use_signals, _Use_watches, _Use_effects, _a;
import { usekey } from "./utils/usekey.js";
import { Counter } from "./utils/counter.js";
import { deep } from "../../../../tools/deep/deep.js";
import { Signal } from "../../../../signals/signal.js";
import { maptool } from "../../../../tools/maptool.js";
import { flat, signals, watch } from "../../../state.js";
export class Use {
    constructor(rerender, context) {
        this[_a] = {
            wrap: (fn) => {
                return ((...args) => {
                    __classPrivateFieldGet(this, _Use_counter, "f").reset();
                    return fn(...args);
                });
            },
            disconnect: () => {
                // cleanup mounts
                for (const down of __classPrivateFieldGet(this, _Use_unmounts, "f"))
                    down();
                __classPrivateFieldGet(this, _Use_unmounts, "f").clear();
                // cleanup inits
                for (const down of __classPrivateFieldGet(this, _Use_initDowns, "f"))
                    down();
                __classPrivateFieldGet(this, _Use_initDowns, "f").clear();
                __classPrivateFieldGet(this, _Use_initResults, "f").clear();
            },
            reconnect: () => {
                // call all mounts
                for (const up of __classPrivateFieldGet(this, _Use_mounts, "f").values())
                    __classPrivateFieldGet(this, _Use_unmounts, "f").add(up());
                // call all inits
                for (const [count, start] of __classPrivateFieldGet(this, _Use_initStarts, "f").entries()) {
                    const [result, down] = start();
                    __classPrivateFieldGet(this, _Use_initResults, "f").set(count, result);
                    __classPrivateFieldGet(this, _Use_initDowns, "f").add(down);
                }
            },
            afterRender: () => {
                for (const [count, fn] of __classPrivateFieldGet(this, _Use_deferred, "f")) {
                    const result = fn();
                    __classPrivateFieldGet(this, _Use_deferredResults, "f").set(count, result);
                }
            },
        };
        _Use_context.set(this, void 0);
        _Use_rerender.set(this, void 0);
        _Use_counter.set(this, new Counter());
        _Use_mounts.set(this, new Map());
        _Use_unmounts.set(this, new Set());
        _Use_initStarts.set(this, new Map());
        _Use_initResults.set(this, new Map());
        _Use_initDowns.set(this, new Set());
        _Use_onces.set(this, new Map());
        _Use_deferred.set(this, new Map());
        _Use_deferredResults.set(this, new Map());
        _Use_states.set(this, new Map());
        _Use_flatstates.set(this, new Map());
        _Use_signals.set(this, new Map());
        _Use_watches.set(this, new Map());
        _Use_effects.set(this, new Map());
        __classPrivateFieldSet(this, _Use_rerender, rerender, "f");
        __classPrivateFieldSet(this, _Use_context, context, "f");
    }
    get context() {
        return __classPrivateFieldGet(this, _Use_context, "f");
    }
    rerender() {
        __classPrivateFieldGet(this, _Use_rerender, "f").call(this);
    }
    mount(fn) {
        const count = __classPrivateFieldGet(this, _Use_counter, "f").pull();
        if (!__classPrivateFieldGet(this, _Use_mounts, "f").has(count)) {
            __classPrivateFieldGet(this, _Use_mounts, "f").set(count, fn);
            __classPrivateFieldGet(this, _Use_unmounts, "f").add(fn());
        }
    }
    init(fn) {
        const count = __classPrivateFieldGet(this, _Use_counter, "f").pull();
        if (!__classPrivateFieldGet(this, _Use_initStarts, "f").has(count)) {
            __classPrivateFieldGet(this, _Use_initStarts, "f").set(count, fn);
            const [result, down] = fn();
            __classPrivateFieldGet(this, _Use_initResults, "f").set(count, result);
            __classPrivateFieldGet(this, _Use_initDowns, "f").add(down);
            return result;
        }
        return __classPrivateFieldGet(this, _Use_initResults, "f").get(count);
    }
    once(prep) {
        const count = __classPrivateFieldGet(this, _Use_counter, "f").pull();
        return maptool(__classPrivateFieldGet(this, _Use_onces, "f")).guarantee(count, prep);
    }
    defer(fn) {
        const count = __classPrivateFieldGet(this, _Use_counter, "f").pull();
        if (!__classPrivateFieldGet(this, _Use_deferred, "f").has(count))
            __classPrivateFieldGet(this, _Use_deferred, "f").set(count, fn);
        return __classPrivateFieldGet(this, _Use_deferredResults, "f").get(count);
    }
    state(init) {
        const count = __classPrivateFieldGet(this, _Use_counter, "f").pull();
        const value = maptool(__classPrivateFieldGet(this, _Use_states, "f")).guarantee(count, () => ((typeof init === "function")
            ? init()
            : init));
        const setter = (v) => {
            __classPrivateFieldGet(this, _Use_states, "f").set(count, v);
            __classPrivateFieldGet(this, _Use_rerender, "f").call(this);
        };
        const getter = () => __classPrivateFieldGet(this, _Use_states, "f").get(count);
        return [value, setter, getter];
    }
    flatstate(init) {
        const count = __classPrivateFieldGet(this, _Use_counter, "f").pull();
        return maptool(__classPrivateFieldGet(this, _Use_flatstates, "f")).guarantee(count, () => (flat.state((typeof init === "function")
            ? init()
            : init)));
    }
    signal(init) {
        const count = __classPrivateFieldGet(this, _Use_counter, "f").pull();
        return maptool(__classPrivateFieldGet(this, _Use_signals, "f")).guarantee(count, () => (signals.signal((typeof init === "function")
            ? init()
            : init)));
    }
    computed(update) {
        const count = __classPrivateFieldGet(this, _Use_counter, "f").pull();
        return maptool(__classPrivateFieldGet(this, _Use_signals, "f")).guarantee(count, () => (signals.computed(update)));
    }
    op() {
        const count = __classPrivateFieldGet(this, _Use_counter, "f").pull();
        return maptool(__classPrivateFieldGet(this, _Use_signals, "f")).guarantee(count, () => signals.op());
    }
    watch(collector) {
        const count = __classPrivateFieldGet(this, _Use_counter, "f").pull();
        return maptool(__classPrivateFieldGet(this, _Use_watches, "f")).guarantee(count, () => watch.track(collector, data => {
            __classPrivateFieldGet(this, _Use_watches, "f").set(count, data);
            __classPrivateFieldGet(this, _Use_rerender, "f").call(this);
        }));
    }
    effect(mount, dependencies) {
        dependencies = dependencies.map(Signal.unwrap);
        const count = __classPrivateFieldGet(this, _Use_counter, "f").pull();
        const effect = __classPrivateFieldGet(this, _Use_effects, "f").get(count);
        if (effect) {
            const [unmount, previous_deps] = effect;
            if (!deep.equal(dependencies, previous_deps)) {
                unmount();
                __classPrivateFieldGet(this, _Use_effects, "f").set(count, [mount(), dependencies]);
            }
        }
        else {
            __classPrivateFieldGet(this, _Use_mounts, "f").set(count, mount);
            const unmount = mount();
            __classPrivateFieldGet(this, _Use_unmounts, "f").add(unmount);
            __classPrivateFieldGet(this, _Use_effects, "f").set(count, [unmount, dependencies]);
        }
    }
}
_Use_context = new WeakMap(), _Use_rerender = new WeakMap(), _Use_counter = new WeakMap(), _Use_mounts = new WeakMap(), _Use_unmounts = new WeakMap(), _Use_initStarts = new WeakMap(), _Use_initResults = new WeakMap(), _Use_initDowns = new WeakMap(), _Use_onces = new WeakMap(), _Use_deferred = new WeakMap(), _Use_deferredResults = new WeakMap(), _Use_states = new WeakMap(), _Use_flatstates = new WeakMap(), _Use_signals = new WeakMap(), _Use_watches = new WeakMap(), _Use_effects = new WeakMap(), _a = usekey;
