// xiome/x/features/store/stripe/liaison/stripe-liaison.js
export function makeStripeLiaison({ stripe }) {
    return {
        accounts: {
            async create(params) {
                return stripe.accounts.create(params);
            },
            async retrieve(id) {
                return stripe.accounts.retrieve(id);
            },
            async createLoginLink(id) {
                return stripe.accounts.createLoginLink(id);
            },
        },
        accountLinks: {
            async create(params) {
                return stripe.accountLinks.create(params);
            },
        },
        account(stripeAccount) {
            const connection = { stripeAccount };
            return {
                customers: {
                    async create(params) {
                        return stripe.customers.create(params, connection);
                    },
                    async retrieve(id) {
                        return stripe.customers.retrieve(id, connection);
                    },
                    async update(id, params) {
                        return stripe.customers.update(id, params, connection);
                    },
                },
                products: {
                    async create(params) {
                        return stripe.products.create(params, connection);
                    },
                    async retrieve(id) {
                        return stripe.products.retrieve(id, connection);
                    },
                    async update(id, params) {
                        return stripe.products.update(id, params, connection);
                    },
                },
                prices: {
                    async create(params) {
                        return stripe.prices.create(params, connection);
                    },
                    async retrieve(id) {
                        return stripe.prices.retrieve(id, connection);
                    },
                    async update(id, params) {
                        return stripe.prices.update(id, params, connection);
                    },
                },
                paymentMethods: {
                    async retrieve(id) {
                        return stripe.paymentMethods.retrieve(id, connection);
                    },
                },
                setupIntents: {
                    async retrieve(id) {
                        return stripe.setupIntents.retrieve(id, connection);
                    },
                },
                subscriptions: {
                    async retrieve(id) {
                        return stripe.subscriptions.retrieve(id, connection);
                    },
                    async update(id, params) {
                        return stripe.subscriptions.update(id, params, connection);
                    },
                },
                checkout: {
                    sessions: {
                        async create(params) {
                            return stripe.checkout.sessions.create(params, connection);
                        },
                    },
                },
            };
        }
    };
}
