// @benev/slate/x/nexus/units/light_view.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
import { render } from "lit";
import { AsyncDirective, directive } from "lit/async-directive.js";
import { UseLightView } from "../parts/use/tailored.js";
import { SlateView } from "../parts/slate_view_element.js";
import { debounce } from "../../tools/debounce/debounce.js";
import { usekey } from "../parts/use/parts/utils/usekey.js";
import { setup_reactivity } from "../parts/setup_reactivity.js";
export const prepare_light_view = ((shell) => (renderer) => {
    var _instances, _props, _element, _render_into_element, _rerender, _use, _rend, _reactivity, _a;
    return directive((_a = class extends AsyncDirective {
            constructor() {
                super(...arguments);
                _instances.add(this);
                _props.set(this, void 0);
                _element.set(this, document.createElement(SlateView.tag));
                _rerender.set(this, debounce(0, () => {
                    if (__classPrivateFieldGet(this, _props, "f") && this.isConnected)
                        this.setValue(__classPrivateFieldGet(this, _instances, "m", _render_into_element).call(this, this.render(...__classPrivateFieldGet(this, _props, "f"))));
                }));
                _use.set(this, new UseLightView(__classPrivateFieldGet(this, _element, "f"), __classPrivateFieldGet(this, _rerender, "f"), shell.context));
                _rend.set(this, __classPrivateFieldGet(this, _use, "f")[usekey].wrap(renderer(__classPrivateFieldGet(this, _use, "f"))));
                _reactivity.set(this, setup_reactivity(__classPrivateFieldGet(this, _rend, "f"), __classPrivateFieldGet(this, _rerender, "f")));
            }
            update(_, props) {
                return __classPrivateFieldGet(this, _instances, "m", _render_into_element).call(this, this.render(...props));
            }
            render(...props) {
                var _b;
                __classPrivateFieldSet(this, _props, props, "f");
                return (_b = __classPrivateFieldGet(this, _reactivity, "f")) === null || _b === void 0 ? void 0 : _b.render(...props);
            }
            reconnected() {
                __classPrivateFieldGet(this, _use, "f")[usekey].reconnect();
                __classPrivateFieldSet(this, _reactivity, setup_reactivity(__classPrivateFieldGet(this, _rend, "f"), __classPrivateFieldGet(this, _rerender, "f")), "f");
            }
            disconnected() {
                __classPrivateFieldGet(this, _use, "f")[usekey].disconnect();
                if (__classPrivateFieldGet(this, _reactivity, "f")) {
                    __classPrivateFieldGet(this, _reactivity, "f").stop();
                    __classPrivateFieldSet(this, _reactivity, undefined, "f");
                }
            }
        },
        _props = new WeakMap(),
        _element = new WeakMap(),
        _rerender = new WeakMap(),
        _use = new WeakMap(),
        _rend = new WeakMap(),
        _reactivity = new WeakMap(),
        _instances = new WeakSet(),
        _render_into_element = function _render_into_element(template) {
            render(template, __classPrivateFieldGet(this, _element, "f"));
            __classPrivateFieldGet(this, _use, "f")[usekey].afterRender();
            return __classPrivateFieldGet(this, _element, "f");
        },
        _a));
});
