// @benev/slate/x/element/gold.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _GoldElement_root, _GoldElement_init, _GoldElement_wait, _GoldElement_render_debounced;
import { render } from "lit";
import { MetallicElement } from "./part/metallic.js";
import { debounce } from "../tools/debounce/debounce.js";
import { explode_promise } from "../tools/explode_promise.js";
import { apply_styles_to_shadow } from "../base/utils/apply_styles_to_shadow.js";
export class GoldElement extends MetallicElement {
    static get styles() { return undefined; }
    init() { }
    constructor() {
        super();
        _GoldElement_root.set(this, void 0);
        _GoldElement_init.set(this, explode_promise());
        _GoldElement_wait.set(this, __classPrivateFieldGet(this, _GoldElement_init, "f").promise);
        _GoldElement_render_debounced.set(this, debounce(0, () => {
            const root = __classPrivateFieldGet(this, _GoldElement_root, "f");
            const template = this.render();
            if (template)
                render(template, root, { host: this });
        }));
        __classPrivateFieldSet(this, _GoldElement_root, this.attachShadow({ mode: "open" }), "f");
        const C = this.constructor;
        apply_styles_to_shadow(__classPrivateFieldGet(this, _GoldElement_root, "f"), C.styles);
        this.init();
    }
    get root() {
        return __classPrivateFieldGet(this, _GoldElement_root, "f");
    }
    get updateComplete() {
        return __classPrivateFieldGet(this, _GoldElement_wait, "f").then(() => true);
    }
    render() { }
    async requestUpdate() {
        const promise = __classPrivateFieldGet(this, _GoldElement_render_debounced, "f").call(this);
        if (__classPrivateFieldGet(this, _GoldElement_init, "f")) {
            promise.then(__classPrivateFieldGet(this, _GoldElement_init, "f").resolve);
            __classPrivateFieldSet(this, _GoldElement_init, undefined, "f");
        }
        __classPrivateFieldSet(this, _GoldElement_wait, promise, "f");
        return promise;
    }
    connectedCallback() {
        super.connectedCallback();
        this.requestUpdate();
    }
}
_GoldElement_root = new WeakMap(), _GoldElement_init = new WeakMap(), _GoldElement_wait = new WeakMap(), _GoldElement_render_debounced = new WeakMap();
